### PR TITLE
Prevent inserting nulls coming from expressions into non nullable columns

### PR DESCRIPTION
**Pull request by homar**
## Description

> Is this change a fix, improvement, new feature, refactoring, or other?

a fix

> Is this a change to the core query engine, a connector, client library, or the SPI interfaces? (be specific)

core query engine

> How would you describe this change to a non-technical end user or system administrator?

it prevents users from inserting nulls into non nullable columns

## Related issues, pull requests, and links
fixes: https://github.com/trinodb/trino/issues/13434
## Documentation

(x) No documentation is needed.
( ) Sufficient documentation is included in this PR.
( ) Documentation PR is available with #prnumber.
( ) Documentation issue #issuenumber is filed, and can be handled later.

## Release notes

( ) No release notes entries required.
(x) Release notes entries required with the following suggested text:

```markdown
# Core
* Prevents inserting nulls that are results of expressions into non nullable columns
```

### CODE REVIEW

File: core/trino-main/src/main/java/io/trino/sql/planner/plan/TableWriterNode.java [hunk 1]
```diff
@@ -55,7 +53,7 @@
     private final Symbol fragmentSymbol;
     private final List<Symbol> columns;
     private final List<String> columnNames;
-    private final Set<Symbol> notNullColumnSymbols;
+    private final List<Symbol> notNullColumnSymbols;
```
Comment (findepi) [inline core/trino-main/src/main/java/io/trino/sql/planner/plan/TableWriterNode.java:56]: Why the change here?
Does the order on this collection matter?

(seems that most of the changes are technical and related to this)

Comment (homar) [reply core/trino-main/src/main/java/io/trino/sql/planner/plan/TableWriterNode.java:56]: So basically I had to make a choice out of 3 options:
1. Change ` map(node.getNotNullColumnSymbols())` in `SymbolMapper` line 406  in such a way that `node.getNotNullColumnSymbols()` is converted to List and then output of this map call is converted back to Set.
2. add another implementation of map function that would accept Set
3. do what I did, also note that other collections here are Lists too

Comment (findepi) [reply core/trino-main/src/main/java/io/trino/sql/planner/plan/TableWriterNode.java:56]: > add another implementation of map function that would accept Set

that would be fine if we could assume `SymbolMapper.map(Symbol)` function is _distinct_.


> 1. Change ` map(node.getNotNullColumnSymbols())` in `SymbolMapper` line 406  in such a way that `node.getNotNullColumnSymbols()` is converted to List and then output of this map call is converted back to Set.

I'd do probably do this.

> also note that other collections here are Lists too

the fields like
```
private final List<Symbol> columns;
private final List<String> columnNames;
```

generally look as describing some fields, so are index-correlated (perhaps)

our `notNullColumnSymbols` doesn't fall into this category, so it's better to be a Set to avoid people assuming some meaning that it doesn't convey.

Comment (findepi) [reply core/trino-main/src/main/java/io/trino/sql/planner/plan/TableWriterNode.java:56]: > 1. Change ` map(node.getNotNullColumnSymbols())` in `SymbolMapper` line 406  in such a way that `node.getNotNullColumnSymbols()` is converted to List and then output of this map call is converted back to Set.

Let's do this (a minimal change) so that we can focus on what the PR actually does and then think what to do.

**Comment by findepi** [comment]
> #13434

If the PR is supposed to fix (close) the issue, prepend the issue number with "fixes: ". Thanks

File: testing/trino-testing/src/main/java/io/trino/testing/BaseConnectorTest.java [hunk 2]
```diff
@@ -2721,6 +2721,7 @@ public void testInsertIntoNotNullColumn()
             assertQuery("SELECT * FROM " + table.getName(), "VALUES (NULL, 2)");
             // This is enforced by the engine and not the connector
             assertQueryFails(format("INSERT INTO %s (not_null_col, nullable_col) VALUES (NULL, 3)", table.getName()), "NULL value not allowed for NOT NULL column: not_null_col");
+            assertQueryFails(format("INSERT INTO %s (not_null_col, nullable_col) VALUES (TRY(5/0), 4)", table.getName()), "NULL value not allowed for NOT NULL column: not_null_col");
```
Comment (findepi) [inline testing/trino-testing/src/main/java/io/trino/testing/BaseConnectorTest.java:2724]: This test passes without the other changes, at least when run with `TestIcebergOrcConnectorTest`.
See https://github.com/trinodb/trino/issues/13434#issuecomment-1203643757 for why this can be the case.

Also, a test case should be also added to all classes overriding this method (clickhouse)

File: core/trino-main/src/main/java/io/trino/sql/planner/optimizations/SymbolMapper.java [hunk 3]
```diff
@@ -403,7 +403,7 @@ public TableWriterNode map(TableWriterNode node, PlanNode source, PlanNodeId new
                 map(node.getFragmentSymbol()),
                 map(node.getColumns()),
                 node.getColumnNames(),
-                node.getNotNullColumnSymbols(),
+                map(ImmutableList.copyOf(node.getNotNullColumnSymbols())).stream().collect(toImmutableSet()),
```
Comment (kasiafi) [inline core/trino-main/src/main/java/io/trino/sql/planner/optimizations/SymbolMapper.java:406]: You can simplify it by using `public Symbol map(Symbol symbol)`

```suggestion
                node.getNotNullColumnSymbols().stream().map(this::map).collect(toImmutableSet()),
```

Comment (homar) [reply core/trino-main/src/main/java/io/trino/sql/planner/optimizations/SymbolMapper.java:406]: but apart from this is it fine in your opinion ?

Comment (kasiafi) [reply core/trino-main/src/main/java/io/trino/sql/planner/optimizations/SymbolMapper.java:406]: The change in `SymbolMapper` is fine. 
Please change the test so that it fails before the change, according to https://github.com/trinodb/trino/pull/13462#discussion_r936383826

Comment (homar) [reply core/trino-main/src/main/java/io/trino/sql/planner/optimizations/SymbolMapper.java:406]: Thank you! Actually I spoke with @findepi and he proposed different approach that I am currently working on.
I will ask for another review when I am done :)

File: core/trino-main/src/main/java/io/trino/sql/planner/LogicalPlanner.java [hunk 4]
```diff
@@ -451,29 +453,39 @@ private RelationPlan getInsertPlan(
...
-                    Expression cast = noTruncationCast(input.toSymbolReference(), queryType, tableType);
-                    assignments.put(output, cast);
+                    expression = noTruncationCast(input.toSymbolReference(), queryType, tableType);
                 }
-                insertedColumnsBuilder.add(column);
             }
+            if (!column.isNullable()) {
+                expression = new IfExpression(
+                        new IsNotNullPredicate(expression),
+                        expression,
```
Comment (findepi) [inline core/trino-main/src/main/java/io/trino/sql/planner/LogicalPlanner.java:481]: I am concerned this can result in `expression` being evaluated twice.
However, we can replace `IF(exp IS NOT NULL, exp, fail(...))` 
with `coalesce(exp, fail(...))`.
Per spec, `coalesce` should evaluate lazily.

Comment (homar) [reply core/trino-main/src/main/java/io/trino/sql/planner/LogicalPlanner.java]: that's smart, thanks

File: core/trino-main/src/main/java/io/trino/sql/planner/LogicalPlanner.java [hunk 5]
```diff
@@ -451,29 +452,36 @@ private RelationPlan getInsertPlan(
...
             int index = insertColumns.indexOf(columns.get(column.getName()));
             if (index < 0) {
                 if (supportsMissingColumnsOnInsert) {
                     continue;
                 }
-                Expression cast = new Cast(new NullLiteral(), toSqlType(column.getType()));
-                assignments.put(output, cast);
-                insertedColumnsBuilder.add(column);
+                if (!column.isNullable()) {
+                    expression = createNullNotAllowedFailExpression(column.getName(), tableType);
```
Comment (findepi) [inline core/trino-main/src/main/java/io/trino/sql/planner/LogicalPlanner.java:460]: Does `createNullNotAllowedFailExpression` get added twice?

File: testing/trino-testing/src/main/java/io/trino/testing/BaseConnectorTest.java [hunk 6]
```diff
@@ -2791,6 +2791,22 @@ public void testInsertIntoNotNullColumn()
             // This is enforced by the engine and not the connector
             assertQueryFails(format("INSERT INTO %s (not_null_col, nullable_col) VALUES (NULL, 3)", table.getName()), "NULL value not allowed for NOT NULL column: not_null_col");
         }
+
+        try (TestTable table = new TestTable(getQueryRunner()::execute, "not_null_no_cast", "(nullable_col INTEGER, not_null_col INTEGER NOT NULL)")) {
```
Comment (findepi) [inline testing/trino-testing/src/main/java/io/trino/testing/BaseConnectorTest.java:2795]: Would it be possible to reuse `TestTable` declaractions?
this would speed up the test a bit

File: testing/trino-testing/src/main/java/io/trino/testing/BaseConnectorTest.java [hunk 7]
```diff
@@ -2791,6 +2791,22 @@ public void testInsertIntoNotNullColumn()
...
+            assertQueryFails(format("INSERT INTO %s (not_null_col, nullable_col) VALUES (NULL, 3)", table.getName()), "NULL value not allowed for NOT NULL column: not_null_col");
+            assertQueryFails(format("INSERT INTO %s (not_null_col, nullable_col) VALUES (TRY(5/0), 4)", table.getName()), "NULL value not allowed for NOT NULL column: not_null_col");
+            assertQueryFails(format("INSERT INTO %s (not_null_col) VALUES (TRY(6/0))", table.getName()), "NULL value not allowed for NOT NULL column: not_null_col");
+        }
+
+        try (TestTable table = new TestTable(getQueryRunner()::execute, "insert_as_select", "(nullable_col INTEGER, not_null_col INTEGER NOT NULL)")) {
+            assertUpdate(format("INSERT INTO %s (not_null_col) VALUES (2)", table.getName()), 1);
+            assertQuery("SELECT * FROM " + table.getName(), "VALUES (NULL, 2)");
+            // This is enforced by the engine and not the connector
+            assertQueryFails(format("INSERT INTO %s (nullable_col) SELECT nationkey from nation", table.getName()), errorMessageForInsertIntoNotNullColumn("not_null_col"));
```
Comment (findepi) [inline testing/trino-testing/src/main/java/io/trino/testing/BaseConnectorTest.java:2808]: uppercase FROM in the query (SQL keyword)

Comment (findepi) [inline testing/trino-testing/src/main/java/io/trino/testing/BaseConnectorTest.java:2808]: Add a test with query inserting implicit NULL (ie non-NULL column not on the insert list) 
from query that returns no rows (eg ... WHER false).

This should succeed. I want this to document current behavior.

Comment (findepi) [inline testing/trino-testing/src/main/java/io/trino/testing/BaseConnectorTest.java:2808]: if this is enforced by the engine, why invoke `errorMessageForInsertIntoNotNullColumn` which would return connector-specific message?

File: core/trino-main/src/main/java/io/trino/sql/planner/LogicalPlanner.java (hunk 5 above)
Comment (homar) [reply core/trino-main/src/main/java/io/trino/sql/planner/LogicalPlanner.java:460]: Depends how you interpret adding twice.
If this one is executed than we end up with `coalesce(createNullNotAllowedFailExpression, createNullNotAllowedFailExpression)`.
Throwing exception from here broke a lot of tests and made it really hard to have a common behaviour between different connectors

File: testing/trino-testing/src/main/java/io/trino/testing/BaseConnectorTest.java (hunk 6 above)
Comment (homar) [reply testing/trino-testing/src/main/java/io/trino/testing/BaseConnectorTest.java:2795]: sure I can give it a try, tough there were 3 already before my change

File: testing/trino-testing/src/main/java/io/trino/testing/BaseConnectorTest.java (hunk 7 above)
Comment (homar) [reply testing/trino-testing/src/main/java/io/trino/testing/BaseConnectorTest.java:2808]: Because there is this if there
`
if (supportsMissingColumnsOnInsert) {
                    continue;
                }
`
and I guess this is expected for some connector, look few lines above where exactly the same method is used in a similar situation

File: core/trino-main/src/main/java/io/trino/sql/planner/LogicalPlanner.java (hunk 5 above)
Comment (findepi) [reply core/trino-main/src/main/java/io/trino/sql/planner/LogicalPlanner.java:460]: > If this one is executed than we end up with `coalesce(createNullNotAllowedFailExpression, createNullNotAllowedFailExpression)`.

That's redundant, and easy to avoid.

File: core/trino-main/src/main/java/io/trino/operator/TableWriterOperator.java [hunk 8]
```diff
@@ -27,7 +27,6 @@
 import io.trino.spi.Mergeable;
 import io.trino.spi.Page;
 import io.trino.spi.PageBuilder;
-import io.trino.spi.TrinoException;
```
Comment (findepi) [inline core/trino-main/src/main/java/io/trino/operator/TableWriterOperator.java:30]: > Remove checking non nullability via TableWriterOperator

Add cmt msg -- "This is now redundant as the checks are inserted by the planner"

File: core/trino-main/src/main/java/io/trino/sql/planner/LogicalPlanner.java (hunk 5 above)
Comment (findepi) [reply core/trino-main/src/main/java/io/trino/sql/planner/LogicalPlanner.java:461]: Why not just

```
expression =  new Cast(new NullLiteral(), toSqlType(column.getType()));
```

?

File: core/trino-main/src/main/java/io/trino/sql/planner/LogicalPlanner.java [hunk 9]
```diff
@@ -451,29 +452,38 @@ private RelationPlan getInsertPlan(
...
                 }
                 else {
-                    Expression cast = noTruncationCast(input.toSymbolReference(), queryType, tableType);
-                    assignments.put(output, cast);
+                    expression = noTruncationCast(input.toSymbolReference(), queryType, tableType);
                 }
-                insertedColumnsBuilder.add(column);
             }
+            if (!column.isNullable() && !missingValueForNotNullableColumn) {
+                expression = new CoalesceExpression(expression, createNullNotAllowedFailExpression(column.getName(), tableType));
```
Comment (findepi) [inline core/trino-main/src/main/java/io/trino/sql/planner/LogicalPlanner.java:480]: why do we need to test `missingValueForNotNullableColumn` here?

File: plugin/trino-mariadb/src/test/java/io/trino/plugin/mariadb/BaseMariaDbConnectorTest.java [hunk 10]
```diff
@@ -262,6 +264,40 @@ public void testDeleteWithLike()
                 .hasStackTraceContaining("TrinoException: Unsupported delete");
     }
 
+    @Test
+    @Override
```
Comment (findepi) [inline plugin/trino-mariadb/src/test/java/io/trino/plugin/mariadb/BaseMariaDbConnectorTest.java:268]: Document  why override.

File: plugin/trino-mariadb/src/test/java/io/trino/plugin/mariadb/BaseMariaDbConnectorTest.java [hunk 11]
```diff
@@ -262,6 +264,40 @@ public void testDeleteWithLike()
...
+            assertQueryFails(
+                    "CREATE TABLE not_null_constraint (not_null_col INTEGER NOT NULL)",
+                    format("line 1:35: Catalog '%s' does not support non-null column for column name 'not_null_col'", getSession().getCatalog().orElseThrow()));
+            return;
+        }
+
+        try (TestTable table = new TestTable(getQueryRunner()::execute, "insert_not_null", "(nullable_col INTEGER, not_null_col INTEGER NOT NULL)")) {
+            assertUpdate(format("INSERT INTO %s (not_null_col) VALUES (2)", table.getName()), 1);
+            assertQuery("SELECT * FROM " + table.getName(), "VALUES (NULL, 2)");
+            // The error message comes from remote databases when ConnectorMetadata.supportsMissingColumnsOnInsert is true
```
Comment (findepi) [inline plugin/trino-mariadb/src/test/java/io/trino/plugin/mariadb/BaseMariaDbConnectorTest.java:283]: in this connector-specific test supportsMissingColumnsOnInsert is known to be true, so the conditional comment ("when") is not applicable

File: plugin/trino-mariadb/src/test/java/io/trino/plugin/mariadb/BaseMariaDbConnectorTest.java [hunk 12]
```diff
@@ -262,6 +264,40 @@ public void testDeleteWithLike()
...
+            assertQueryFails(format("INSERT INTO %s (nullable_col) VALUES (1)", table.getName()), errorMessageForInsertIntoNotNullColumn("not_null_col"));
+            assertQueryFails(format("INSERT INTO %s (not_null_col, nullable_col) VALUES (NULL, 3)", table.getName()), "NULL value not allowed for NOT NULL column: not_null_col");
+            assertQueryFails(format("INSERT INTO %s (not_null_col, nullable_col) VALUES (TRY(5/0), 4)", table.getName()), "NULL value not allowed for NOT NULL column: not_null_col");
+            assertQueryFails(format("INSERT INTO %s (not_null_col) VALUES (TRY(6/0))", table.getName()), "NULL value not allowed for NOT NULL column: not_null_col");
+            assertQueryFails(format("INSERT INTO %s (nullable_col) SELECT nationkey FROM nation", table.getName()), errorMessageForInsertIntoNotNullColumn("not_null_col"));
+            // TODO  this query should not fail, when this is fixed this method doesn't need to be overriden here https://github.com/trinodb/trino/issues/13551
```
Comment (findepi) [inline plugin/trino-mariadb/src/test/java/io/trino/plugin/mariadb/BaseMariaDbConnectorTest.java:289]: ( https://github.com/trinodb/trino/issues/13551 )

```
// TODO (https://github.com/trinodb/trino/issues/13551) This doesn't fail for other connectors so
//  probably shouldn't fail for MariaDB either. Once fixed, remove test override.
```

(mind two spaces after `//` on the second comment line)

File: core/trino-main/src/main/java/io/trino/sql/planner/LogicalPlanner.java (hunk 9 above)
Comment (homar) [reply core/trino-main/src/main/java/io/trino/sql/planner/LogicalPlanner.java:480]: To make sure that we didn't already set it here:
`
if (!column.isNullable()) {
                    expression = createNullNotAllowedFailExpression(column.getName(), tableType);
                    missingValueForNotNullableColumn = true;
                }
`

Comment (findepi) [reply core/trino-main/src/main/java/io/trino/sql/planner/LogicalPlanner.java:480]: right, but the above could be simplified -- https://github.com/trinodb/trino/pull/13462#discussion_r941093005

Comment (homar) [reply core/trino-main/src/main/java/io/trino/sql/planner/LogicalPlanner.java:480]: so I thought this way it is more efficient, it will fail on the first expression in the coercion, but sure I can change it

File: plugin/trino-mariadb/src/test/java/io/trino/plugin/mariadb/BaseMariaDbConnectorTest.java [hunk 13]
```diff
@@ -262,6 +264,41 @@ public void testDeleteWithLike()
...
+    {
+        skipTestUnless(hasBehavior(SUPPORTS_CREATE_TABLE));
+
+        if (!hasBehavior(SUPPORTS_NOT_NULL_CONSTRAINT)) {
+            assertQueryFails(
+                    "CREATE TABLE not_null_constraint (not_null_col INTEGER NOT NULL)",
+                    format("line 1:35: Catalog '%s' does not support non-null column for column name 'not_null_col'", getSession().getCatalog().orElseThrow()));
+            return;
+        }
+
```
Comment (findepi) [inline plugin/trino-mariadb/src/test/java/io/trino/plugin/mariadb/BaseMariaDbConnectorTest.java:280]: Remove. It's known whether MariaDB has specific behavior.

**Review by findepi — APPROVED** [review]

File: plugin/trino-mariadb/src/test/java/io/trino/plugin/mariadb/BaseMariaDbConnectorTest.java [hunk 14]
```diff
@@ -262,6 +262,32 @@ public void testDeleteWithLike()
...
+            assertQueryFails(format("INSERT INTO %s (not_null_col, nullable_col) VALUES (NULL, 3)", table.getName()), "NULL value not allowed for NOT NULL column: not_null_col");
+            assertQueryFails(format("INSERT INTO %s (not_null_col, nullable_col) VALUES (TRY(5/0), 4)", table.getName()), "NULL value not allowed for NOT NULL column: not_null_col");
+            assertQueryFails(format("INSERT INTO %s (not_null_col) VALUES (TRY(6/0))", table.getName()), "NULL value not allowed for NOT NULL column: not_null_col");
+            assertQueryFails(format("INSERT INTO %s (nullable_col) SELECT nationkey FROM nation", table.getName()), errorMessageForInsertIntoNotNullColumn("not_null_col"));
+            // TODO (https://github.com/trinodb/trino/issues/13551) This doesn't fail for other connectors so
+            //  probably shouldn't fail for MariaDB either. Once fixed, remove test override.
+            assertQueryFails(format("INSERT INTO %s (nullable_col) SELECT nationkey FROM nation WHERE regionkey < 0", table.getName()), ".*Field 'not_null_col' doesn't have a default value.*");
```
Comment (findinpath) [inline plugin/trino-mariadb/src/test/java/io/trino/plugin/mariadb/BaseMariaDbConnectorTest.java:280]: There seems to be a problem here caused by the current implementation


I modified the test to run like this:
```
    public void testInsertIntoNotNullColumn()
    {
        try (TestTable table = new TestTable(getQueryRunner()::execute, "insert_not_null", "(nullable_col INTEGER, not_null_col INTEGER NOT NULL)")) {
            assertQuerySucceeds(format("INSERT INTO %s (nullable_col) SELECT nationkey FROM nation WHERE regionkey < 0", table.getName()));
        }
```

and I notice the following exception in the stacktrace:

```
java.lang.AssertionError: Expected query to succeed: INSERT INTO insert_not_null1evdhkhr9c (nullable_col) SELECT nationkey FROM nation WHERE regionkey < 0
...
    Suppressed: java.lang.Exception: SQL: INSERT INTO insert_not_null1evdhkhr9c (nullable_col) SELECT nationkey FROM nation WHERE regionkey < 0
.........
         Suppressed: java.lang.RuntimeException: Query: INSERT INTO `tpch`.`insert_not_null1evdhkhr9c` (`nullable_col`) SELECT `nullable_col` FROM `tpch`.`tmp_trino_2dc66901e1fa421f9cf3064225e9df13
```


There seems to be some side-effect of your changes.

[the remark this answers is not in the document]
Comment (homar) [reply plugin/trino-mariadb/src/test/java/io/trino/plugin/mariadb/BaseMariaDbConnectorTest.java:280]: that won't work here

**Comment by homar** [comment]
Failure same as here https://github.com/trinodb/trino/issues/13556

File: plugin/trino-mariadb/src/test/java/io/trino/plugin/mariadb/BaseMariaDbConnectorTest.java (hunk 14 above)
Comment (findinpath) [reply plugin/trino-mariadb/src/test/java/io/trino/plugin/mariadb/BaseMariaDbConnectorTest.java:280]: I get it now. The exception originates from MariaDB which complaints while copying the content from the temporary table towards the target table.

**Review by findepi — APPROVED** [review]

**Comment by homar** [comment]
raptor failure is a know flaky https://github.com/trinodb/trino/issues/12385

**Comment by homar** [comment]
Second failure is `io.trino.tests.product.iceberg.TestIcebergSparkCompatibility.testStatsAfterAddingPartitionField` which is not a known flaky but root cause seems like a transient issue:
`
Caused by: org.apache.hadoop.ipc.RemoteException(java.io.IOException): File /user/hive/warehouse/test_stats_after_adding_partition_field_co9knj4i0fkp/data/col1=4/00000-2016-70242f10-85f4-447a-ab96-8776b4233d2e-00001.parquet could only be written to 0 of the 1 minReplication nodes. There are 1 datanode(s) running and no node(s) are excluded in this operation.
`

**Comment by homar** [comment]
trino-hive failure seems not to be related https://github.com/trinodb/trino/issues/13600

**Comment by homar** [comment]
In default, suite-7-non-generic  `TestIcebergSparkCompatibility.testTrinoReadsTrinoTableWithSparkDeletesAfterOptimizeAndCleanUp` failed which is not a known issue but root cause seems to be a transient issue(like the one above - maybe there is some hidden problem somewhere?) 
`
 Caused by: org.apache.hadoop.ipc.RemoteException(java.io.IOException): File /user/hive/warehouse/test_spark_reads_trino_partitioned_table_with_deletes_after_expiring_snapshots_after_optimizeparquet-55279a4a1bd84f34b032479abbb27345/metadata/00003-fa053566-1af5-424f-af09-befb86799553.metadata.json could only be written to 0 of the 1 minReplication nodes. There are 1 datanode(s) running and no node(s) are excluded in this operation.
`

**Comment by findepi** [comment]
`TestIcebergSparkCompatibility` failed twice in a row (at different methods) with `could only be written to 0 of the 1 minReplication nodes. There are 1 datanode(s) running and no node(s) are excluded in this operation.`.
We have some retries for that in Hive product tests. 
Can this be applied in this class too?

**Comment by homar** [comment]
> `TestIcebergSparkCompatibility` failed twice in a row (at different methods) with `could only be written to 0 of the 1 minReplication nodes. There are 1 datanode(s) running and no node(s) are excluded in this operation.`. We have some retries for that in Hive product tests. Can this be applied in this class too?

I think it can be. I will create a ticket for it

**Comment by homar** [comment]
> `TestIcebergSparkCompatibility` failed twice in a row (at different methods) with `could only be written to 0 of the 1 minReplication nodes. There are 1 datanode(s) running and no node(s) are excluded in this operation.`. We have some retries for that in Hive product tests. Can this be applied in this class too?

https://github.com/trinodb/trino/issues/13601
